source/Calamari/ArgoCD/Conventions/UpdateGitRepositoryInstallConvention.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            this.fileSystem = fileSystem;
+            this.repositoryParentDirectory = repositoryParentDirectory;
+            this.log = log;
+        }
+        
+        public void Install(RunningDeployment deployment)
+        {
+            var repositoryIndexes = deployment.Variables.GetIndexes(SpecialVariables.Git.Index);
+
+            Log.Info("Executing Commit To Git operation");
+            var fileGlob = deployment.Variables.GetPaths(SpecialVariables.Git.TemplateGlobs);
+            var filesToApply = SelectFiles(deployment.CurrentDirectory, fileGlob).ToList();
+            
+            var commitMessage = GenerateCommitMessage(deployment);
+            var requiresPullRequest = FeatureToggle.ArgocdCreatePullRequestFeatureToggle.IsEnabled(deployment.Variables) && deployment.Variables.Get(SpecialVariables.Git.CommitMethod) == "PullRequest";
+            
+            log.Info($"Found {filesToApply.Count} files to apply");
+            
+            var repositoryFactory = new RepositoryFactory(log, repositoryParentDirectory);
+            var repositoryIndexNames = repositoryIndexes.Join(",");
+            log.Info($"Found the following repository indicies '{repositoryIndexNames}'");
+            foreach (var repositoryIndex in repositoryIndexes)
+            {
+                Log.Info($"Writing files to repository for index {repositoryIndex}");
+                IGitConnection gitConnection = new VariableBackedGitConnection(deployment.Variables, repositoryIndex);
+
+                var repository = repositoryFactory.CloneRepository(repositoryIndex, gitConnection);
+                Log.Info("Copying files into repository");
+                var filesAdded = CopyFilesIntoPlace(filesToApply, repository.WorkingDirectory, gitConnection.SubFolder);
+                Log.Info("Staging files in repository");
+                repository.StageFiles(filesAdded);
+                Log.Info("Commiting changes");
+                if (repository.CommitChanges(commitMessage))
+                {
+                    Log.Info("Changes were commited, pushing to remote");
+                    repository.PushChanges(requiresPullRequest, gitConnection.BranchName);    
+                }
+                else
+                {
+                    Log.Info("No changes were commited.");
+                }
+            }
+        }
+
+        string GenerateCommitMessage(RunningDeployment deployment)
+        {
+            var summary = deployment.Variables.GetMandatoryVariable(SpecialVariables.Git.CommitMessageSummary);
+            var description = deployment.Variables.Get(SpecialVariables.Git.CommitMessageDescription) ?? string.Empty;
+            return description.Equals(string.Empty)
+                ? summary
+                : $"{summary}\n\n{description}";
+        }
+
+
+        List<string> CopyFilesIntoPlace(IEnumerable<FileToCopy> filesToCopy, string destinationRootDir, string repoSubFolder)
```

Comment:
Any thoughts on testing this function separately? Or happy to test as part of `UpdateGitRepositoryInstallConventionTests`

Reply:
yeah - my tests are goinv to tackle each bit incrementally I think

Reply:
I've split it up but don't love it